aws-http-auth/sigv4/sigv4.go

```diff
@@ -0,0 +1,154 @@
+// Package sigv4 implements request signing for the basic form AWS Signature
+// Version 4.
+package sigv4
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/smithy-go/aws-http-auth/credentials"
+	v4internal "github.com/aws/smithy-go/aws-http-auth/internal/v4"
+	v4 "github.com/aws/smithy-go/aws-http-auth/v4"
+)
+
+const algorithm = "AWS4-HMAC-SHA256"
+
+// Signer signs requests with AWS Signature version 4.
+type Signer struct {
+	options v4.SignerOptions
+}
+
+// New returns an instance of Signer with applied options.
+func New(opts ...v4.SignerOption) *Signer {
+	options := v4.SignerOptions{}
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &Signer{options}
+}
+
+// SignRequestInput is the set of inputs for Sigv4 signing.
+type SignRequestInput struct {
+	// The input request, which will modified in-place during signing.
+	Request *http.Request
+
+	// The SHA256 hash of the input request body.
+	//
+	// This value is NOT required to sign the request, but it is recommended to
+	// provide it (or provide a Body on the HTTP request that implements
+	// io.Seeker such that the signer can calculate it for you). Many services
+	// do not accept requests with unsigned payloads.
+	//
+	// If a value is not provided, and DisableImplicitPayloadHashing has not
+	// been set on SignerOptions, the signer will attempt to derive the payload
+	// hash itself. The request's Body MUST implement io.Seeker in order to do
+	// this, if it does not, the magic value for unsigned payload is used. If
+	// the body does implement io.Seeker, but a call to Seek returns an error,
+	// the signer will forward that error.
+	PayloadHash []byte
+
+	// The identity used to sign the request.
+	Credentials credentials.Credentials
+
+	// The service and region for which this request is to be signed.
+	//
+	// The appropriate values for these fields are determined by the service
+	// vendor.
+	Service, Region string
```

Comment:
That's entirely up to the authorizer. So for AWS, it can of course only be our regions. But some custom service with its own authorizer could use whatever values here.